### PR TITLE
feat(aws-lambda-elasticachmemcached): New Construct

### LINKED ISSUE #674: feat(aws-lambda-elasticachmemcached): New Construct
Create a new construct that links a new or existing cache construct to a new or existing Lambda function construct.

### PR DESCRIPTION
**Pull request by biffgaut**
*Issue #, if available:*
Closes #674 

*Description of changes:*
Create a new construct that links a new or existing cache construct to a new or existing Lambda function construct.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 402b58acfad80490180c42ba8cffed04528b733e
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c16e2650-604a-4413-84df-3bf08be9521d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 402b58acfad80490180c42ba8cffed04528b733e
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3504e378-e751-4e37-8cfb-8d2acc2ede2d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+|existingVpc?|[`ec2.IVpc`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.IVpc.html)|An optional, existing VPC into which this pattern should be deployed. When deployed in a VPC, the Lambda function will use ENIs in the VPC to access network resources and an Interface Endpoint will be created in the VPC for Amazon SQS. If an existing VPC is provided, the `deployVpc` property cannot be `true`. This uses `ec2.IVpc` to allow clients to supply VPCs that exist outside the stack using the [`ec2.Vpc.fromLookup()`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.Vpc.html#static-fromwbrlookupscope-id-options) method.|
+|vpcProps?|[`ec2.VpcProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.VpcProps.html)|Optional user provided properties to override the default properties for the new VPC. `subnetConfiguration` is set by the pattern, so any values for those properties supplied here will be overrriden. |
+| cacheEndpointEnvironmentVariableName?| string | |
```
Comment (mickychetta) [inline source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:59]: Missing description for cacheEndpointEnvironmentVariableName

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+Python
+```python
+
+```
+
+Java
+``` java
+
+```
```
Comment (mickychetta) [inline source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:49]: Missing Python and Java minimal deployments

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+|vpcProps?|[`ec2.VpcProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.VpcProps.html)|Optional user provided properties to override the default properties for the new VPC. `subnetConfiguration` is set by the pattern, so any values for those properties supplied here will be overrriden. |
+| cacheEndpointEnvironmentVariableName?| string | |
+| cacheProps? | [`cache.CfnCacheClusterProps`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticache.CfnCacheClusterProps.html) | Optional user provided props to override the default props for the Elasticache Cluster. Providing both this and `existingCache` will cause an error.' |
```
Comment (mickychetta) [inline source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:60]: Extra ending single quote at the end of description

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+* Configure limited privilege access IAM role for Lambda function
+* Enable reusing connections with Keep-Alive for NodeJs Lambda function
+* Enable X-Ray Tracing
+* Attached to self referencing security group to grant access to cache
+* Set Environment Variables
+  * (default) CACHE_ENDPOINT
+  * AWS_NODEJS_CONNECTION_REUSE_ENABLED (for Node 10.x and higher functions)
+
+### Amazon Elasticache Memcached Cluster
+* Creates multi node, cross-az cluster by default
```
Comment (mickychetta) [inline source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:85]: Should we mention default cacheNodeType and numCacheNodes

**Review by mickychetta — CHANGES_REQUESTED** [review]
Architecture is good, just minor documentation comments.

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md (hunk 2 above)
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:49]: Yep, put those off for later :-)

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md (hunk 4 above)
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/README.md:85]: Good

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: d687bce1891b98aa2d9611a4fcd68f647516bb03
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=45ff3518-02df-48a8-af83-8753f0393590%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: d687bce1891b98aa2d9611a4fcd68f647516bb03
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ca1f8d56-1a39-482e-b952-59163b60d4bb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 5d1cb53e4a335ecc64168ac0a36ec0c7e6288815
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b554e87b-eab2-4271-999b-84727866248a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 5d1cb53e4a335ecc64168ac0a36ec0c7e6288815
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3648ffd8-559b-4d50-8f30-fb07944db717%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: d350a31f2f3eeff2bcfe0b508491c12ab7fccd89
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=24debf65-79e5-4147-a8c0-b8b1a1ad4816%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: d350a31f2f3eeff2bcfe0b508491c12ab7fccd89
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4e0c6b6f-5dbd-4898-8498-c9906fd98e2b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 9b9d7f5508d386516e281163447883789312b93d
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6701bbb5-23d7-4d7b-a211-809c1eb3870e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 9b9d7f5508d386516e281163447883789312b93d
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3269b79e-674f-4ad8-be01-886b49932cbf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 3e7d517121c244ea367a5e48c22240680dfd56db
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2c8f31a7-90d0-496f-8b4a-563b2985869a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 19f90542ef63d188780a37703878fb0620cc8c8f
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3d12e96e-5ec4-4bba-8c01-70aff22604fd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: b3babf78f4b90c13b4cfe2545178cd08938710dc
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ccf25772-a72c-478a-9f8b-be3a924c32eb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: b3babf78f4b90c13b4cfe2545178cd08938710dc
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f9ea1bf0-ca68-4d2d-906f-8f5f58d8e710%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by mickychetta — CHANGES_REQUESTED** [review]
Overall good, just minor documentations and rearranging of functions. Tests were very easy to read and clear.

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/lib/index.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    if (
+      props.cacheProps &&
+      props.cacheProps.engine &&
+      props.cacheProps.engine !== "memcached"
+    ) {
+      throw Error("This construct can only launch memcached clusters");
+    }
+    if (props.cacheProps && props.existingCache) {
+      throw Error("Cannot specify existingCache and cacheProps");
+    }
```
[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/lib/index.ts:104]: Yep, leaving that for you do when you write fargate-elasticachememcached :-)

[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/lib/index.ts:104]: I think there's a few others that can move at that time too.

File: source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/lib/index.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      defaultEnvironmentVariableName,
+      props.cacheEndpointEnvironmentVariableName
+    );
+  }
+}
+
+function AddLambdaEnvironmentVariable(targetFunction: lambda.Function, value: string, defaultName: string, clientName?: string) {
+  const variableName = clientName || defaultName;
+  targetFunction.addEnvironment(variableName, value);
+}
```
[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-lambda-elasticachememcached/lib/index.ts:151]: That was my original thought. But looking at the code left behind to call it I'm not sure it's an improvement - figured we should talk it over.

File: source/patterns/@aws-solutions-constructs/core/lib/security-group-helper.ts [hunk 7]
```diff
@@ -55,3 +55,36 @@ export function buildSecurityGroup(
...
+      reason: "Egress of 0.0.0.0/0 is default and generally considered OK",
+    },
+    {
+      id: "W40",
+      reason:
+        "Egress IPProtocol of -1 is default and generally considered OK",
+    },
+  ]);
+  return newCacheSG;
+}
```
[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/core/lib/security-group-helper.ts:90]: I agree - but let's wait until we have 2 examples before we try to infer the generic.

File: source/patterns/@aws-solutions-constructs/core/test/elasticache-defaults.test.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+});
+
+test("Test GetMemcachedDefaults()", () => {
+  const testPort = 22222;
+  const testId = 'test';
+
+  const props = GetMemcachedDefaults(testId, testPort);
+
+  expect(props.port).toEqual(testPort);
+  expect(props.clusterName).toEqual(`${testId}-cdk-cluster`);
```
[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/core/test/elasticache-defaults.test.ts:30]: My thought was to ensure that using 1 static value and 1 dynamic value ensured that the functionality was returning the right mix, but I will add these. My concern is that the more redundant data we have between this code and elasticache-defaults the more brittle the test.

File: source/patterns/@aws-solutions-constructs/core/test/test-helper.ts [hunk 9]
```diff
@@ -106,4 +110,37 @@ export function suppressAutoDeleteHandlerWarnings(stack: Stack) {
...
+
+  // Create the subnet group from all the isolated subnets in the VPC
+  const subnetGroup = CreateCacheSubnetGroup(scope, vpc, id);
+  const emptySG = new ec2.SecurityGroup(scope, `${id}-cachesg`, {
+    vpc,
+    allowAllOutbound: true,
+  });
+  addCfnSuppressRules(emptySG, [{ id: "W40", reason: "Test Resource" }]);
+  addCfnSuppressRules(emptySG, [{ id: "W5", reason: "Test Resource" }]);
+  addCfnSuppressRules(emptySG, [{ id: "W36", reason: "Test Resource" }]);
```
[the remark this answers is not in the document]
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/core/test/test-helper.ts:126]: I don't think it's that common. It appears twice in the actual application in the two functions in security-group-helper.ts, here (a shared location) and in one of my new integration tests. 

I think I can change my integration test to use the security group helper so it only appears in shared locations.

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 2348453e76aa539aacdd44f0a93d3b18a2e03f34
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bcfaad35-73ce-4055-9fd8-3adc1c7126a1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 2348453e76aa539aacdd44f0a93d3b18a2e03f34
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1d3af256-c64f-41ee-a5da-7ca01855e2c4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by mickychetta — APPROVED** [review]

**Pull request merged**